ydb/library/yql/minikql/computation/mkql_vector_spiller_adapter.h

```diff
@@ -0,0 +1,207 @@
+            default:
+                return;
+        }
+    }
+
+    ///Get requested vector.
+    std::vector<T>&& ExtractVector() {
+        StoredChunksElementsCount.pop();
+        State = EState::AcceptingDataRequests;
+        return std::move(CurrentVector);
+    }
+
+    ///Start restoring next vector. If th eentire contents of the vector are in memory
+    ///State will be changed to DataREady without any async read operation. ExtractVector is expected
+    ///to be called immediately.
+    void RequestNextVector() {
+        MKQL_ENSURE(State == EState::AcceptingDataRequests, "Internal logic error");
+        MKQL_ENSURE(CurrentVector.empty(), "Internal logic error");
+        MKQL_ENSURE(!StoredChunksElementsCount.empty(), "Internal logic error");
+
+        CurrentVector.reserve(StoredChunksElementsCount.front());
+        State = EState::RestoringData;
+
+        LoadNextVector();
+    }
+
+    ///Finalize will spill all the contents of inner buffer if any.
+    ///Is case if buffer is not ready async write operation will be started.
+    void Finalize() {
+        MKQL_ENSURE(CurrentVector.empty(), "Internal logic error");
+        if (Buffer.empty()) {
+            State = EState::AcceptingDataRequests;
+            return;
+        }
+
+        SaveBuffer();
+        IsFinalizing = true;
+    }
+
+private:
+
+    void CopyRopeToTheEndOfVector(std::vector<T>& vec, TRope& rope) {
+        for (auto it = rope.begin(); it != rope.end(); ++it) {
+            const T* data = reinterpret_cast<const T*>(it.ContiguousData());
+            vec.insert(vec.end(), data, data + it.ContiguousSize() / sizeof(T)); // size is always multiple of sizeof(T)
+        }
+    }
+
+    void LoadNextVector() {
+        auto requestedVectorSize= StoredChunksElementsCount.front();
+        MKQL_ENSURE(requestedVectorSize >= CurrentVector.size(), "Internal logic error");
+        size_t sizeToLoad = (requestedVectorSize - CurrentVector.size()) * sizeof(T);
+
+        if (Buffer.size() >= sizeToLoad) {
+            // if all the data for requested vector is ready
+            TRope remainingPartOfVector = Buffer.Extract(Buffer.Position(0), Buffer.Position(sizeToLoad));
+            CopyRopeToTheEndOfVector(CurrentVector, remainingPartOfVector);
+            State = EState::DataReady;
+        } else {
+            CopyRopeToTheEndOfVector(CurrentVector, Buffer);
+            ReadOperation = Spiller->Extract(StoredChunks.front());
+        }
+    }
+
+    void SaveBuffer() {
+        State = EState::SpillingData;
+        WriteOperation = Spiller->Put(std::move(Buffer));
+    }
+
+    void AddDataToRope(T* data, size_t count) {
+        Buffer.Insert(Buffer.End(), TRope(TString(reinterpret_cast<const char*>(data), count * sizeof(T))));
+    }
+
+    void SaveNextPartOfVector() {
+        size_t maxFittingElemets = (SizeLimit - Buffer.size()) / sizeof(T);
+        size_t remainingElementsInVector = CurrentVector.size() - NextVectorPositionToSave;
```

Comment:
can we get negative values here?

Reply:
No, this function is the only place where `NextVectorPositionToSave` is changed. Initially the value is zero and when the value reaches `CurrentVector.size()` it resets to zero. 174 line of current file